### PR TITLE
fix(import): Set sequences after import

### LINKED ISSUE #3840: When adding a new project after reimporting old projects, fails due to postgres sequence not having been reset
## Situation

I maintain a set of docker containers for developers to use.  As much as possible, these are plug-and-play - so a user just runs `docker-compose up` and gets a full stack with all requisite services pre-configured and databases loaded with seed data.  For sentry's part, I simply use `sentry dump` and `sentry import` to have sentry's dataset ready to go - which for the most part works flipping great.
## Setup to repeat

Set up a sentry installation in docker and add a project.
### sentry.json (initial)

``` json
[]
```
### Dockerfile

``` dockerfile
FROM sentry:8.5

ADD docker-entrypoint.sh /entrypoint.sh
ADD sentry.json /sentry.json

USER sentry
```
### docker-entrypoint.sh

``` bash
#!/bin/bash

set -e

# first check if we're passing flags, if so
# prepend with sentry
if [ "${1:0:1}" = '-' ]; then
    set -- sentry "$@"
fi

case "$1" in
    start)
        echo "Migrations starting..."

        sleep 15
        sentry upgrade --noinput || (echo "Sentry upgrade failed"; exit 1)
        sentry createuser "--email=dev@nlpsecure.com" "--password=test123" "--superuser" "--no-input" || echo "Sentry createuser failed - may already be there"

        echo "Migrations completed (hopefully)"

        echo "Running import"
        sentry import /sentry.json
        echo "Done running import"

        set -- sentry "$@"
    ;;
    run|celery|cleanup|config|createuser|devserver|django|export|help|import|init|plugins|queues|repair|shell)
        echo "Waiting 120 seconds for database to be ready..."
        sleep 120
        echo "Starting sentry workers..."

        set -- sentry "$@"
    ;;
    generate-secret-key)
        exec sentry config generate-secret-key
    ;;
esac

exec "$@"
```
### docker-compose.yml:

``` yaml
redis:
    container_name: redis
    image: redis
    labels:
        SERVICE_ID: redis
    command: redis-server --appendonly yes
    dns: 172.20.254.2
    dns_search: docker
    ports:
        - "6379:6379"

postgres:
    container_name: postgres
    image: postgres
    labels:
        SERVICE_ID: postgres
    environment:
        POSTGRES_PASSWORD: secret
        POSTGRES_USER: sentry
    restart: always

# Sentry and sentry workers - error tracking

sentry:
    container_name: sentry
    image: gcr.io/nlp-development-tools/sentry:latest
    ports:
        - "9000:9000"
    labels:
        SERVICE_ID: sentry
    command: "start"
    links:
        - "redis:redis"
        - "postgres:postgres"
    environment:
        SENTRY_POSTGRES_PORT: "5432"
        SENTRY_REDIS_PORT: "6379"
        SENTRY_SECRET_KEY: "sldakfjalsdkfjlasdkfjasldkfj"
    restart: always

sentry-celery-beat:
    container_name: sentry-celery-beat
    image: gcr.io/nlp-development-tools/sentry:latest
    links:
        - "redis:redis"
        - "postgres:postgres"
    labels:
        SERVICE_ID: sentry-celery-beat
    command: "run cron"
    environment:
        SENTRY_POSTGRES_PORT: "5432"
        SENTRY_REDIS_PORT: "6379"
        SENTRY_SECRET_KEY: "alsdfjkasldkfjlasdkfj"
    restart: always

sentry-celery-worker1:
    container_name: sentry-celery-worker1
    image: gcr.io/nlp-development-tools/sentry:latest
    links:
        - "redis:redis"
        - "postgres:postgres"
    labels:
        SERVICE_ID: sentry-celery-worker
    command: "run worker"
    environment:
        SENTRY_POSTGRES_PORT: "5432"
        SENTRY_REDIS_PORT: "6379"
        SENTRY_SECRET_KEY: "asldkfjasldkfjasldkfjlasdkfj"
    restart: always
```
#### Sentry export created with:

``` bash
#!/bin/bash

docker exec -it sentry sentry export -q > sentry.json

cat sentry.json | jq
```
## Run and add a project and let workers start up.

Run an export to sentry.json so you get your whole project, and then restart.  It should just come online with all your projects, users, etc.  Awesome!
## Issue

After having done this process, there are several situations in which sentry runs into primary key conflicts.  It looks like Postgres is restarting its sequences at primary key 1, while Sentry's import has already inserted ID=1.  The easiest to duplicate is to just add a new project when there's already one in place.  Workers also dump some info about primary key conflicts into their error output and internal reporting during their startup/checkin.
## Possible solutions

I'm likely less familiar with Postgres than you guys, but it seems that just iterating through all the affected tables during import and adjusting the current sequence number to the max value would resolve it.

http://stackoverflow.com/questions/244243/how-to-reset-postgres-primary-key-sequence-when-it-falls-out-of-sync
## Eternal gratefulness

Thanks for all that you guys do, and I love your product!  I'm slowly and surely working The Management towards purchasing a fully supported on-premise copy.

### LINKED ISSUE #23843: I have an issue when I import the export json file
## Important Details

On-Premise w/ Docker, version 9.1.2

## Description

I have two servers with a sentry. There are several projects on the first server, and I would like to copy these projects to the clear second server.
I use the export/import commands. The export command works fine. However, when I run the import command, I get an error.

## Steps to Reproduce

1. Run command 'sentry export sentry_export.json' on the first server
2. Run command 'sentry import sentry_export.json' on the second server
3. Get an error

Good items to include here include:

`Traceback (most recent call last):
  File "/usr/local/bin/sentry", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/site-packages/sentry/runner/__init__.py", line 162, in main
    cli(prog_name=get_prog(), obj={}, max_content_width=100)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sentry/runner/decorators.py", line 36, in inner
    return ctx.invoke(f, *args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/sentry/runner/commands/backup.py", line 21, in import_
    for obj in serializers.deserialize("json", src, stream=True, use_natural_keys=True):
  File "/usr/local/lib/python2.7/site-packages/django/core/serializers/json.py", line 76, in Deserializer
    six.reraise(DeserializationError, DeserializationError(e), sys.exc_info()[2])
  File "/usr/local/lib/python2.7/site-packages/django/core/serializers/json.py", line 70, in Deserializer
    for obj in PythonDeserializer(objects, **options):
  File "/usr/local/lib/python2.7/site-packages/django/core/serializers/python.py", line 140, in Deserializer
    data[field.name] = field.to_python(field_value)
  File "/usr/local/lib/python2.7/site-packages/sentry/db/models/fields/array.py", line 56, in to_python
    value = json.loads(value)
  File "/usr/local/lib/python2.7/site-packages/sentry/utils/json.py", line 111, in loads
    return _default_decoder.decode(value)
  File "/usr/local/lib/python2.7/site-packages/simplejson/decoder.py", line 370, in decode
    obj, end = self.raw_decode(s)
  File "/usr/local/lib/python2.7/site-packages/simplejson/decoder.py", line 400, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
django.core.serializers.base.DeserializationError: Expecting value: line 1 column 2 (char 1)`

### What you expected to happen

Import works fine too

### Possible Solution

[If you have an idea on how this could be solved include that detail here.]

### PR DESCRIPTION
**Pull request by BYK**
Fixes #3840.

### CODE REVIEW

File: src/sentry/runner/commands/backup.py [hunk 1]
```diff
@@ -8,11 +8,22 @@
...
 
     for obj in serializers.deserialize("json", src, stream=True, use_natural_keys=True):
         obj.save()
 
+    sequence_reset_sql = StringIO()
+
+    for app in settings.INSTALLED_APPS:
+        management.call_command("sqlsequencereset", app, stdout=sequence_reset_sql)
+
+    from django.db import connection
```
Comment (wedamija) [inline src/sentry/runner/commands/backup.py:23]: You could probably put this and all the other imports at the top of the file

File: src/sentry/runner/commands/backup.py [hunk 2]
```diff
@@ -8,11 +8,22 @@
...
+        with connection.cursor() as cursor:
+            cursor.execute(sequence_reset_sql.getvalue())
```
Comment (wedamija) [inline src/sentry/runner/commands/backup.py:25]: bad indent

**Comment by BYK** [comment]
@Agalin - would you mind trying this build to see if it fixes the issues you are having: `SENTRY_IMAGE=us.gcr.io/sentryio/sentry:dd0f70c1f1df4cc1c2e683b9b2d018a91ef37208`

**Comment by Agalin** [comment]
I'll check it tomorrow when I get to the office.

**Comment by Agalin** [comment]
Doesn't affect initial import, seems to be a separate issue where Content Types and other tables already have the same data that is in the backup after running just `install.sh`. Can't check if it fixes sequence issues I've observed after importing as it raises exception.
```
# docker-compose run --rm -T web import /etc/sentry/backup.json
Starting sentry_onpremise_symbolicator_1 ... done
Starting sentry_onpremise_clickhouse_1   ... done
Starting sentry_onpremise_zookeeper_1    ... done
Starting sentry_onpremise_redis_1        ... done
Starting sentry_onpremise_smtp_1         ... done
Starting sentry_onpremise_kafka_1        ... done
Starting sentry_onpremise_snuba-consumer_1              ... done
Starting sentry_onpremise_snuba-api_1                   ... done
Starting sentry_onpremise_snuba-replacer_1              ... done
Starting sentry_onpremise_snuba-outcomes-consumer_1     ... done
Starting sentry_onpremise_snuba-sessions-consumer_1     ... done
08:46:49 [INFO] sentry.plugins.github: apps-not-configured
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/apps/registry.py", line 149, in get_app_config
    return self.app_configs[app_label]
KeyError: 'django.contrib.admin'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 471, in handle
    app_configs = [apps.get_app_config(app_label) for app_label in app_labels]
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 471, in <listcomp>
    app_configs = [apps.get_app_config(app_label) for app_label in app_labels]
  File "/usr/local/lib/python3.6/site-packages/django/apps/registry.py", line 156, in get_app_config
    raise LookupError(message)
LookupError: No installed app with label 'django.contrib.admin'. Did you mean 'admin'?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/sentry", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/sentry/runner/__init__.py", line 164, in main
    cli(prog_name=get_prog(), obj={}, max_content_width=100)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/sentry/runner/decorators.py", line 28, in inner
    return ctx.invoke(f, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/sentry/runner/commands/backup.py", line 24, in import_
    management.call_command("sqlsequencereset", app, stdout=sequence_reset_sql)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/__init__.py", line 131, in call_command
    return command.execute(*args, **defaults)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.6/site-packages/django/core/management/base.py", line 473, in handle
    raise CommandError("%s. Are you sure your INSTALLED_APPS setting is correct?" % e)
django.core.management.base.CommandError: No installed app with label 'django.contrib.admin'. Did you mean 'admin'?. Are you sure your INSTALLED_APPS setting is correct?
```

**Comment by BYK** [comment]
@Agalin the exception you got should be fixed now. I'll try this later myself but if you can (or want to) you can try it with `SENTRY_IMAGE=us.gcr.io/sentryio/sentry:f53e33c32570b979f0573a2493b5dc45afa1a798`

**Comment by Agalin** [comment]
Seems to be fine in the db. Checked for repositories that definitely had wrong sequence number before.

**Review by wedamija — APPROVED** [review]

**Comment by BYK** [comment]
Locally verified too, merging.

**Pull request merged**